File: HW/HW8/LongestPalindrome.java

import java.util.*;
class LongestPalindrome {

    public int longestPalindrome(String s){
        if ((1 <= s.length() && s.length()<= 2000)) {
            int i, count = 0;
            char ch = ' ';
            String s_new = "", s_palin1 ="", s_palin2 ="";
            for(i = 0; i< s.length(); i++){
                ch = s.charAt(i);
                if(Character.isLetter(ch)){
                    s_new = s_new + ch;
                }
                else{
                    return -1;
                }
            }
            HashMap<Character, Integer> map_Palin = new HashMap<>();

            for ( i = 0; i<s_new.length(); i++){
                ch = s_new.charAt(i);
                if(map_Palin.containsKey(ch)){
                    map_Palin.put(ch, map_Palin.get(ch) + 1);
                }
                else
                    map_Palin.put(ch,1);
            }

            for(Map.Entry<Character, Integer> e : map_Palin.entrySet()){
                i = e.getValue()/2;
                for(int j = 0; j<i;j++){
                    s_palin1 = s_palin1 + e.getKey();
                    s_palin2 = e.getKey() + s_palin2;
                    count++;
                }
                if(s_palin2 == "" && (e.getValue()%2)!= 0){
                    ch = e.getKey();
                    count++;
                }
            }

            return (s_palin1+ch+s_palin2).length();
          }
        else{
            return -1;
        }
    
    }
    public static void main (String args[]){
        Scanner in = new Scanner (System.in);

        LongestPalindrome palinOb = new LongestPalindrome();
        System.out.println("Enter the phrase");
        String s = in.nextLine();
        int result = palinOb.longestPalindrome(s);
        if (result !=-1)
            System.out.println(result);
        else
            System.out.println("Invalid Entry");

    }
    
    
}

File: HW/HW8/LongestPalindrome.py
class LongestPalindrome:
    def longestPalindrome(self, s) -> int:
        if 1 <= len(s) <= 2000 and s.isalpha():
            map_Palin = {}
            s_palin1 = ""
            s_palin2 = ""
            ch = ''
            for ch in s:
                if ch in map_Palin:
                    map_Palin[ch] += 1
                else:
                    map_Palin[ch] = 1
            ch = ' '
            for key, value in map_Palin.items():
                i = value // 2
                for j in range(i):
                    s_palin1 += key
                    s_palin2 = key + s_palin2
                if ch == ' ' and value % 2 != 0:
                    ch = key
            if ch !=' ':
                return len(s_palin1 + ch + s_palin2)
            else:
                return len(s_palin1 + s_palin2)

        else:
            return -1

    def main(self):
        s = input('Enter your first input:')
        result = self.longestPalindrome(s)
        if result != -1:
            print(result)
        else:
            print("Invalid Entry")


if __name__ == "__main__":
    palinOb = LongestPalindrome()
    palinOb.main()
